File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        def _dfs(node):
            if not node: return None
            if node.val == p.val or node.val == q.val: 
                return node 
            
            left = _dfs(node.left)
            right = _dfs(node.right)
            
            if left and right:
                return node
            elif left:
                return left
            else:
                return right

        return _dfs(root)

File: 236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    
    const _dfs = (node) => {
        if (!node) return null;
        if(node.val === p.val || node.val === q.val) return node;

        const left = _dfs(node.left);
        const right = _dfs(node.right);

        if (left && right) {
            return node;
        } else if (left) {
            return left;
        } else {
            return right;
        }
    }
    return _dfs(root);
};